docs/components/security.rst

```diff
@@ -1,23 +1,365 @@
+
+Mautic takes the summation of the bits for the permissions given to a Role and stores it in the database. For example, if a Role has view and edit access, the stored bit is 3. If it has view and create access, the stored bit is 5.
+
+When validating permissions, for example ``plugin:helloWorld:worlds:create``, Mautic checks if Role's generated bit has at least 4 for ``plugin:helloWorld:worlds``. If so, permission gets granted.
+
+.. note:: The permission ``full`` grants access to all previous permissions within the level and thus should always be the highest bit.
+
+Using permissions
+*****************
+
+You can use permissions as follows in your controllers and services:
+
+.. code-block:: php
+
+   <?php
+   declare(strict_types=1);
+
+   /** @var \Mautic\CoreBundle\Security\Permissions\CorePermissions */
+   $security = $this->get('mautic.security');
+
+   // Check if user is granted a single permission
+   if ($security->isGranted('plugin:helloWorld:worlds:view')) {
+       // do something
+   }
+
+   // Check if user is granted multiple permissions (must be granted to all to be true)
+   if ($security->isGranted(
+       array(
+           'plugin:helloWorld:worlds:view',
+           'plugin:helloWorld:worlds:create',
+       )
+   )
+   ) {
+       //do something
+   }
+
+   // Check if user is granted to at least one permission
+   if ($security->isGranted(
+       array(
+           'plugin:helloWorld:worlds:view',
+           'plugin:helloWorld:worlds:edit',
+       ),
+       'MATCH_ONE'
+   )
+   ) {
+       //do something
+   }
+
+   // Get an array of user permissions
+   $permissions = $security->isGranted(
+       array(
+           'plugin:helloWorld:worlds:view',
+           'plugin:helloWorld:worlds:edit',
+       ),
+       'RETURN_ARRAY'
+   );
+
+   if ($permissions['plugin:helloWorld:worlds:view']) {
+       // do something
+   }
+
+   // Check if user has access to view leads
+   if ($security->isGranted('lead:leads:viewother')) {
+       // do something
+   }
+
+To determine if a User has a specific permission, use Mautic's security service which has the ``mautic.security`` alias. The class is ``Mautic\CoreBundle\Security\Permissions\CorePermissions``.
+
+As suggested in the preceding section, Mautic uses a special permission notation to refer to a specific permission.
+For core bundles, use ``bundleName:permissionLevel:permission``.
+For Plugins, append ``plugin:``, for example ``plugin:bundleName:permissionLevel:permission``. ``plugin:`` tells Mautic to look for the permission class in the ``plugins/`` directory and the ``MauticPlugin`` namespace. 
+
+Mautic the core bundle or Plugin to set permission levels and permissions. For example, Mautic's core UserBundle has ``users`` and ``roles`` levels with ``view``, ``edit``, ``create``, ``delete`` and ``full`` permissions for each. 
```

Comment:
This first sentence seems incomplete - do you mean you should use the core bundle or plugin to set the permission/levels?

Reply:
Good catch! Updated:

> Mautic's core bundles and external Plugins are responsible for setting the permission level and permissions. For example, Mautic's core UserBundle has ``users`` and ``roles`` levels with ``view``, ``edit``, ``create``, ``delete`` and ``full`` permissions for each.